### PR TITLE
Add graph support for strategic test models

### LINKED ISSUE #1077: Support Testable flow Phase 1
### Goal

Support `testable` flow in studio including test mode and revamping of service and mapping tests editors. 

### Problem
We want to address the inconsistency of writing service and mapping tests as well as writing test data.
Additionally it is not very intuitive to create service tests today. To address this we will  move from a `Lambda` based assertion to specific type of assertions such as EqualToJSON etc.
Additionally the introduction of `DataElement` will allow to more intuitive create test data and share data across multiple tests. 
Furthermore when executing test via studio the new apis will give more context to better debug failing tests such as execution result etc. 

### Phase 1: Graph and basic editor support
- ability to define following testables with the following test types and embedded data types in text mode
- ability to run the relevant tests through the global test runner (Legacy Test will not be supported via this mode) -> https://github.com/finos/legend-studio/issues/865. 
- Testables will run as part of the projects pipeline flow
- ensure both flows use the testable test runners in engine. Studio should not use them to run
- Testables to Support: `Service`, `Mapping`
- Assert Types: `EqualTo`, `EqualToJSON`, `EqualToTDS`
- Embedded Data: `Relational`, `ServiceStore` , `ExternalFormat` `ModelStore`
 
### Phase 2: Advanced editor support (to be moved to a separate issue)
- Service Testable Editor 
- Mapping Testable Editor
- Editors for the following Embedded Data Types: `Relational`, `ServiceStore` and `ExternalFormat`
- Enhance form support for the embedded data type editors by allowing users to specify the store while creating the data element and auto populate the properties for embedded data

### Phase 3: Migration Plan from Legacy (to be moved to a separate issue)
_TBD_

## Implementation
In order to fully support the new testable flow we must ensure the following 
- [ ] [pure](https://github.com/finos/legend-pure)/[engine](https://github.com/finos/legend-engine) Current testable elements tests are updated to the new testable framework
  - [ ] `Mapping` @gs-gunjan 
  - [x] `Service` @hardikmaheshwari  
- [ ]  [pure](https://github.com/finos/legend-pure)/[engine](https://github.com/finos/legend-engine) `DataElements` are updated to support the follow embedded data
    - [x] `ServiceStore` @hardikmaheshwari 
    - [ ] `FlatData` - need to add a test to ensure we can migrate flat data tests to the new testable flow
    - [ ] `Relational` @MauricioUyaguari
      - [x] Add Relational Data Compiler, Grammar, Studio
      - [x] Add `equalToTDS` Compiler,Grammar, Studio
      - [x] Relational should work with equalToJSON as well
      - [ ]  Ensure `equalToTDS` tests run with api
      - [ ] `Investigate` relational queries that return graph fetch break while running tests. (feeling its because the return type is complex [firm with person properties]
    - [x] `ExternalFormat` @hardikmaheshwari 
    - [x] `ModelStore` @hardikmaheshwari 
- [ ] [engine](https://github.com/finos/legend-engine) Ensure testable elements `tests` are run as part of run testable apis
  - [ ] `Mapping` @gs-gunjan 
  - [x] `Service` @hardikmaheshwari 
- [ ] [sdlc](https://github.com/finos/legend-sdlc) Ensure testable elements `tests` are run in sdlc pipeline @hardikmaheshwari 
- [ ] Support `DataElement` in text mode
    - [x] `ServiceStore` ( https://github.com/finos/legend-studio/issues/1078) 
    - [x] `Relational` 
    - [ ] `Flatdata`, still need ? 
    - [x] `ExternalFormat`  
    - [x] `ModelStore` 
- [ ] Support Testable `tests` in text mode
  - [ ] `Mapping` 
  - [x] `Service` 
 - [ ] Support `DataElement` in form mode
    - [x] `ServiceStore` 
    - [ ] `Relational` 
    - [x] `ExternalFormat`  #1133 
    - [ ] `ModelStore` 
- [x] Support executing Testable
- [x] Rework Service Test Editor
- [ ] Rework Mapping Test Editor
- [ ] Figure out Strategy for dealing with legacy tests. i.e transform them to new, read but not write. etc. 
- [x] V0 of global test runner 
  - [x] Add panel of global test runner
  - [x] run tests for all testables 

### Concerns/To Check
- [ ] should `RelationalTestData` support sql statements as input
- [ ] Base 64 Encoding of TestData to avoid escape hacks

### Data Element Model
image

### Test and Test Result Model
image

### Relevant PRS:
- [ ] https://github.com/finos/legend-pure/pull/544
- [ ] https://github.com/finos/legend-pure/pull/533
- [x] https://github.com/finos/legend-pure/pull/521
- [ ] https://github.com/finos/legend-engine/pull/658
- [x] https://github.com/finos/legend-engine/pull/631
- [x] https://github.com/finos/legend-sdlc/pull/418
- [x] https://github.com/finos/legend-studio/pull/1114

### WIP

Relational Test Grammar
```
###Data
Data my::RelationalData
{
  Relational
  #{
    PersonTable:
      {
        columns:
        [
          'id',
          'firstName',
          'lastName'
        ];
        rows:
        [
          '1,Mauricio,Uyaguari',
          '2,An,Phi'
        ];
      }
  }#
}


###Service
Service model::MyService
{
  pattern: '/d2c48a9c-70fa-46e3-8173-c355e774004f';
  documentation: '';
  autoActivateUpdates: true;
  execution: Single
  {
    query: |model::Person.all()->project([x|$x.firstName, x|$x.lastName], ['First Name', 'Last Name']);
    mapping: mapping::RelationalMapping;
    runtime: model::Runtime;
  }
  testSuites:
  [
    testSuite1:
    {
      data:
      [
        connections:
        [
          connection_1:
            Reference
            #{
              my::RelationalData
            }#
        ]
      ]
      tests:
      [
        test1:
        {
          asserts:
          [
            assert1:
              EqualToTDS
              #{
                expected : 
                  {
                    columns: 
                    [
                      'firstName',
                      'lastName'
                    ];
                    rows:
                    [
                      '1,Mauricio,Uyaguari',
                      '2,An,Phi'
                    ];
                  };
              }#
          ]
        }
      ]
    }
  ]
  test: Single
  {
    data: 'default\nPersonTable\nid,firstName,lastName\n1,John,Doe\n2,An,Phi\n3,Mauricio,Uyaguari\n-----\n';
    asserts:
    [
      { [], res: meta::pure::mapping::Result[1]|$res.values->cast(@TabularDataSet)->toOne().rows->map(r|$r.values->makeString(',')) == ['John,Doe', 'An,Phi', 'Mauricio,Uyaguari'] }
    ];
  }
}


###Relational
Database store::TestDB
(
  Table PersonTable
  (
    id INTEGER PRIMARY KEY,
    firm_id INTEGER,
    firstName VARCHAR(200),
    lastName VARCHAR(200)
  )
)


###Pure
Class model::Person
{
  firstName: String[1];
  lastName: String[1];
}


###Mapping
Mapping mapping::RelationalMapping
(
  *model::Person: Relational
  {
    ~primaryKey
    (
      [store::TestDB]PersonTable.id
    )
    ~mainTable [store::TestDB]PersonTable
    firstName: [store::TestDB]PersonTable.firstName,
    lastName: [store::TestDB]PersonTable.lastName
  }

  MappingTests
  [
    test_csv
    (
      query: |model::Person.all()->project([x|$x.firstName, x|$x.lastName], ['First Name', 'Last Name']);
      data:
      [
        <Relational, CSV, store::TestDB, 
          'default\n'+
          'PersonTable\n'+
          'id,firstName,lastName\n'+
          '1,Mauricio,Uyaguari\n'+
          '2,An,Phi\n'+
          '\n\n\n'
        >
      ];
      assert: '[{"values":["Mauricio","Uyaguari"]},{"values":["An","Phi"]}]';
    )
  ]
)


###Connection
RelationalDatabaseConnection model::MyConnection
{
  store: store::TestDB;
  type: H2;
  specification: LocalH2
  {
    testDataSetupSqls: [
      'Drop table if exists FirmTable;\nDrop table if exists PersonTable;\nCreate Table FirmTable(id INT, Legal_Name VARCHAR(200));\nCreate Table PersonTable(id INT, firm_id INT, lastName VARCHAR(200), firstName VARCHAR(200));\nInsert into FirmTable (id, Legal_Name) values (1, \'FirmA\');\nInsert into FirmTable (id, Legal_Name) values (2, \'Apple\');\nInsert into PersonTable (id, firm_id, lastName, firstName) values (1, 1, \'John\', \'Doe\');\nInsert into PersonTable (id, firm_id, lastName, firstName) values (2, 2, \'Tim\', \'Smith\');\nInsert into PersonTable (id, firm_id, lastName, firstName) values (3, 3, \'Nicole\', \'Doe\');\n\n'
      ];
  };
  auth: DefaultH2;
}


###Runtime
Runtime model::Runtime
{
  mappings:
  [
    mapping::RelationalMapping
  ];
  connections:
  [
    store::TestDB:
    [
      connection_1: model::MyConnection
    ]
  ];
}

```

### LINKED ISSUE #1078: Feature request: Support modelling of ServiceStore EmbeddedData
### Similar requests

- [X] I have searched and found no existing similar requests

### How are you using Studio?

_No response_

### What problems are you trying to solve?

Support modelling of ServiceStore EmbeddedData

### Describe the solution you would like to see implemented

- [x] Add graph support for `ServiceStoreEmbeddedData`
- [ ] Add form support for `ServiceStoreEmbeddedData`

### Describe alternatives you have considered

_No response_

### Documentation, Design, Adoption, Migration Strategy

_No response_

### Contribution

- [X] I would like to work on this feature

### PR DESCRIPTION
**Pull request by gayathrir11**
## Summary

Closes #1077 
Closes #1078 

## How did you test this change?

- [x] Test(s) added
- [ ] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)
image

image

### CODE REVIEW

File: packages/legend-graph/src/models/protocols/pure/v1/model/test/assertion/V1_EqualTo.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+export class V1_EqualTo extends V1_TestAssertion implements Hashable {
+  expected!: object;
+
+  get hashCode(): string {
+    return hashArray([
+      CORE_HASH_STRUCTURE.EQUAL_TO,
+      this.id,
+      hashObjectWithoutSourceInformation(this.expected),
+    ]); //add this.expected
```
Comment (akphi) [inline packages/legend-graph/src/models/protocols/pure/v1/model/test/assertion/V1_EqualTo.ts:30]: why are we not doing this already?

File: packages/legend-graph/src/models/metamodels/pure/packageableElements/service/ServiceTest_Legacy.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import { type Hashable, hashArray, uuid } from '@finos/legend-shared';
+import { CORE_HASH_STRUCTURE } from '../../../../../MetaModelConst';
+import type { RawLambda } from '../../rawValueSpecification/RawLambda';
+import type { Service } from './Service';
+
+/**
+ * TODO: Remove once migration from `ServiceTest_Legacy` to `ServiceTest` is complete
+ * @deprecated
+ */
+export abstract class ServiceTest_Legacy implements Hashable {
```
Comment (akphi) [inline packages/legend-graph/src/models/metamodels/pure/packageableElements/service/ServiceTest_Legacy.ts:26]: Can we make the effort to change this to `DEPREACTED__ServiceTest`. Thanks!

Might need to apply this to my other comments.

File: packages/legend-graph/src/models/metamodels/pure/packageableElements/service/ServiceTest_Legacy.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  abstract get hashCode(): string;
+}
+
+/**
+ * TODO: Remove once migration from `ServiceTest_Legacy` to `ServiceTest` is complete
+ * @deprecated
+ */
+export class TestContainer implements Hashable {
```
Comment (akphi) [inline packages/legend-graph/src/models/metamodels/pure/packageableElements/service/ServiceTest_Legacy.ts:40]: Let's also add `DEPREACTED__` to these models

File: packages/legend-graph/src/models/metamodels/pure/packageableElements/service/ServiceTest_Legacy.ts (hunk 2 above)
Comment (akphi) [reply packages/legend-graph/src/models/metamodels/pure/packageableElements/service/ServiceTest_Legacy.ts:26]: Also remember to change the name of the files

File: packages/legend-graph/src/models/protocols/pure/v1/model/packageableElements/service/V1_ServiceTest_Legacy.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+import { type Hashable, hashArray } from '@finos/legend-shared';
+import { CORE_HASH_STRUCTURE } from '../../../../../../../MetaModelConst';
+import type { V1_RawLambda } from '../../rawValueSpecification/V1_RawLambda';
+
+/**
+ * TODO: Remove once migration from `V1_ServiceTest_Legacy` to `V1_ServiceTest` is complete
+ * @deprecated
+ */
+export abstract class V1_ServiceTest_Legacy implements Hashable {
```
Comment (akphi) [inline packages/legend-graph/src/models/protocols/pure/v1/model/packageableElements/service/V1_ServiceTest_Legacy.ts:25]: Please update the name of this model and this file

File: packages/legend-graph/src/graphManager/action/changeDetection/DSLService_ObserverHelper.ts [hunk 5]
```diff
@@ -98,7 +185,9 @@ export const observe_MultiExecutionTest = skipObserved(
   },
 );
 
-export const observe_ServiceTest = (metamodel: ServiceTest): ServiceTest => {
+export const observe_ServiceTest_Legacy = (
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/action/changeDetection/DSLService_ObserverHelper.ts:188]: though we rename these models to `DEPRECATED__...`, I think we can keep calling these methods `observe_LegacyServiceTest` for now. What do you think?

File: packages/legend-graph/src/models/metamodels/pure/packageableElements/service/TestData.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/**
```
Comment (akphi) [inline packages/legend-graph/src/models/metamodels/pure/packageableElements/service/TestData.ts:1]: Let's name this `ServiceTestData`

File: packages/legend-graph/src/graphManager/action/changeDetection/TestObserverHelper.ts [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+      hashCode: computed,
+    });
+
+    observe_AtomicTestId(metamodel.atomicTestId);
+
+    return metamodel;
+  },
+);
+
+export function observe_AtomicTest(metamodel: AtomicTest): AtomicTest {
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/action/changeDetection/TestObserverHelper.ts:186]: Shouldn't this method take a context as well?

File: packages/legend-extension-external-store-service/src/graphManager/action/changeDetection/ESService_ObserverHelper.ts [hunk 8]
```diff
@@ -380,3 +388,90 @@ export const observe_RootServiceInstanceSetImplementation =
...
+}
+
+export const observe_ServiceRequestPattern = skipObserved(
+  (metamodel: ServiceRequestPattern): ServiceRequestPattern => {
+    makeObservable(metamodel, {
+      url: observable,
+      urlPath: observable,
+      method: observable,
+      headerParams: observable,
+      queryParams: observable,
```
Comment (akphi) [inline packages/legend-extension-external-store-service/src/graphManager/action/changeDetection/ESService_ObserverHelper.ts:432]: why don't we observe these maps?

File: packages/legend-graph/src/graphManager/action/changeDetection/DataObserverHelper.ts [hunk 9]
```diff
@@ -0,0 +1,112 @@
+/**
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/action/changeDetection/DataObserverHelper.ts:1]: --> `DSLData_ObserverHelper`?

File: packages/legend-graph/src/graphManager/action/changeDetection/TestObserverHelper.ts [hunk 10]
```diff
@@ -0,0 +1,210 @@
+/**
```
Comment (akphi) [inline packages/legend-graph/src/graphManager/action/changeDetection/TestObserverHelper.ts:1]: Let's rename this to `Test_ObserverHelper`

**Review by akphi — CHANGES_REQUESTED** [review]
There are a lot of cover. I don't think I have fully inspected all of the observers though 😢 But yeah, just left minor comments.

File: packages/legend-graph/src/graphManager/action/changeDetection/TestObserverHelper.ts (hunk 7 above)
Comment (MauricioUyaguari) [reply packages/legend-graph/src/graphManager/action/changeDetection/TestObserverHelper.ts:186]: We don't need that as of now. We will need it once we add stuff like AssertEqualTDS

File: packages/legend-graph/src/models/protocols/pure/v1/model/test/assertion/V1_EqualTo.ts (hunk 1 above)
Comment (MauricioUyaguari) [reply packages/legend-graph/src/models/protocols/pure/v1/model/test/assertion/V1_EqualTo.ts:30]: already added, removed the comment.

File: packages/legend-graph/src/models/protocols/pure/v1/V1_PureGraphManager.ts [hunk 11]
```diff
@@ -816,6 +821,11 @@ export class V1_PureGraphManager extends AbstractPureGraphManager {
     await this.buildServices(graph, inputs, options);
     stopWatch.record(GRAPH_MANAGER_EVENT.GRAPH_BUILDER_SERVICES_BUILT);
 
+    //build data elements
```
[the remark this answers is not in the document]
Comment (MauricioUyaguari) [reply packages/legend-graph/src/models/protocols/pure/v1/V1_PureGraphManager.ts:824]: as discussed, data elements go after

**Review by akphi — APPROVED** [review]

**Pull request merged**